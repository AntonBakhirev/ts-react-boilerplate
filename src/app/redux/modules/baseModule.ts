import {Action} from "redux";

export interface IAction<P, T = string, M = string> extends Action<T> {
  payload?: P;
  message?: M;
}

export interface IBaseState {
  error: string;
  loaded: boolean;
  pending: boolean;
}

// trick compiler that they are literal types
export type INVOKED = "INVOKED";
export type PENDING = "PENDING";
export type FULFILLED = "FULFILLED";
export type REJECTED = "REJECTED";

export interface IPromiseActions {
  INVOKED: INVOKED;
  PENDING: PENDING;
  FULFILLED: FULFILLED;
  REJECTED: REJECTED;
}

export function getPromiseActionType(baseAction: string): IPromiseActions {
  return {
    FULFILLED: baseAction + "_FULFILLED" as FULFILLED,
    INVOKED: baseAction as INVOKED,
    PENDING: baseAction + "_PENDING" as PENDING,
    REJECTED: baseAction + "_REJECTED" as REJECTED
  };
}

export interface IActionCreators<I, P, F, R> {
  actionTypes: IPromiseActions;
  invoke: (payload: I) => IAction<I, INVOKED>;
  setPending: (payload: P) => IAction<P, PENDING>;
  setFulfilled: (payload: F) => IAction<F, FULFILLED>;
  setRejected: (payload: R) => IAction<null, REJECTED, R>;
}

export function getPromiseAction<I, P, F, R>(type: string): IActionCreators<I, P, F, R> {
  return {
    actionTypes: getPromiseActionType(type),
    invoke: (payload: I): IAction<I, INVOKED> => {
      return {type: type as INVOKED, payload};
    },
    setFulfilled: (payload: F): IAction<F, FULFILLED> => {
      return {type: getPromiseActionType(type).FULFILLED, payload};
    },
    setPending: (payload: P): IAction<P, PENDING> => {
      return {type: getPromiseActionType(type).PENDING, payload};
    },
    setRejected: (message: R): IAction<null, REJECTED, R> => {
      return {type: getPromiseActionType(type).REJECTED, message};
    }
  };
}
