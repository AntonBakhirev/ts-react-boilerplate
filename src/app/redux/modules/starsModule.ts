import {IStars, IStarsAction} from "models/starsModel";
import promiseReducer from "../../helpers/promiseReducer";

/** Action Types */
export const STARS_REQUEST: string = "stars/STARS_REQUEST";
export const STARS_SUCCESS: string = "stars/STARS_SUCCESS";
export const STARS_FAILURE: string = "stars/STARS_FAILURE";

/** Initial State */
const initialState: IStars = {
  isFetching: false
};

/** Reducer */
export function starsReducer(state: IStars = initialState, action: IStarsAction): IStars {
  return promiseReducer<IStars, IStarsAction>("stars/STARS", state, action);
}
