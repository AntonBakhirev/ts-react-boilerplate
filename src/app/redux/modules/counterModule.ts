import {IAction, IState} from "./baseModule";

export const INCREMENT: string = "counter/INCREMENT";
export const DECREMENT: string = "counter/DECREMENT";

export interface ICounter {
  count: number;
}

/** Counter: Initial State */
const initialState: IState<ICounter> = {
  isFetching: false,
  payload: {
    count: 0
  }
};

/** Reducer: CounterReducer */
export function counterReducer(state: IState<ICounter> = initialState, action?: IAction<ICounter>): IState<ICounter> {
  switch (action.type) {
    case INCREMENT:
      return {
        ...state,
        payload: {
          count: state.payload.count + 1
        }
      };

    case DECREMENT:
      return {
        ...state,
        payload: {
          count: ((state.payload.count - 1 > 0) ? state.payload.count - 1 : 0)
        }
      };

    default:
      return state;
  }
}
