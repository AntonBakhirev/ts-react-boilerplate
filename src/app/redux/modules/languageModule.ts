import promiseReducer, {IRequestType} from "../../helpers/promiseReducer";
import {IAction, IState} from "./baseModule";

export const SET_LANGUAGE = "language/SET_LANGUAGE";
export const SWITCH_LANGUAGE = "language/SWITCH_LANGUAGE";
export const requestType: IRequestType = {
  FAILURE: "language/FAILURE",
  PENDING: "language/PENDING",
  SUCCESS: SET_LANGUAGE
};

export interface ILanguage {
  languageData: any;
  locale: string;
}

const initialState: IState<ILanguage> = {
  isFetching: false,
  payload: {
    languageData: {},
    locale: "en-GB"
  }
};

export function languageReducer(state: IState<ILanguage> = initialState, action: IAction<ILanguage>): IState<ILanguage> {
  return promiseReducer<ILanguage>(requestType, state, action);
}
