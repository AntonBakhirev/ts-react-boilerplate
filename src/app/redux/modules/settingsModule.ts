import {FULFILLED, getPromiseAction, IAction, IBaseState, INVOKED, PENDING, REJECTED} from "./baseModule";

export const changeLanguage = getPromiseAction<string, null, any, string>("settings/CHANGE_LANGUAGE");

export interface ITranslations {
  [key: string]: string;
}

export interface ISettingsState extends IBaseState {
  language: string;
  translations: ITranslations;
}

const initialState: ISettingsState = {
  error: "",
  language: "en-US",
  pending: false,
  translations: {}
};

type CHANGE_LANGUAGE_PENDING = PENDING;
type CHANGE_LANGUAGE_FULFILLED = FULFILLED;
type CHANGE_LANGUAGE_REJECTED = REJECTED;

export function settingsReducer(
  state: ISettingsState = initialState,
  action: IAction<null, CHANGE_LANGUAGE_PENDING> |
    IAction<ITranslations, CHANGE_LANGUAGE_FULFILLED> |
    IAction<null, CHANGE_LANGUAGE_REJECTED, string>
): ISettingsState {
  switch (action.type) {
    case changeLanguage.actionTypes.PENDING:
      return {
        ...state,
        pending: true
      };
    case changeLanguage.actionTypes.FULFILLED:
      return {
        ...state,
        translations: action.payload
      };
    case changeLanguage.actionTypes.REJECTED:
      return {
        ...state,
        error: action.message,
        pending: false
      };
    default:
      return state;
  }
}

export function invokeChangeLanguage(language: string): IAction<string, INVOKED> {
  return {
    payload: language,
    type: changeLanguage.actionTypes.INVOKED
  };
}
