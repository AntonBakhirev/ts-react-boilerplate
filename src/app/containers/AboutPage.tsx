import * as React from "react";
import {connect} from "react-redux";
import {Dispatch} from "redux";
import {createSelector} from "reselect";
import {oc} from "../helpers/oc";
import {Translator} from "../models/Translator";
import {ITranslator} from "../models/TranslatorInterfaces";
import {IStore} from "../redux/IStore";
import {invokeChangeLanguage} from "../redux/modules/settingsModule";

class AboutPage extends React.Component<IStateToProps & IDispatchToProps> {
  constructor(props: IStateToProps & IDispatchToProps) {
    super(props);
    this.switchLanguage = this.switchLanguage.bind(this);
  }

  public switchLanguage(): void {
    const language = this.props.language === "en-US" ? "de" : "en-US";
    this.props.invokeChangeLanguage(language);
  }

  public render(): JSX.Element {
    const {language, translations: {aboutUs, change, currentLanguage}} = this.props;
    return (
      <div>
        <h3>{aboutUs}: {language}</h3>
        <button onClick={this.switchLanguage}>
          {change}
        </button>
        <h4>{currentLanguage}</h4>
      </div>
    );
  }
}

interface IStateToProps {
  language: string;
  translations: {
    aboutUs: string;
    change: string;
    currentLanguage: string;
  };
}

interface IDispatchToProps {
  invokeChangeLanguage: (language: string) => void;
}

const translationsSelector = (state: Pick<IStore, "settings">) => oc(state).settings.translations();

const componentTranslationsSelector = createSelector(
  translationsSelector,
  (translations) => {
    const translator: ITranslator = new Translator(translations);
    return {
      aboutUs: translator.translate("about.us"),
      change: translator.translate("about.change"),
      currentLanguage: translator.translate("current.language")
    };
  }
);

export const mapStateToProps = (state: Pick<IStore, "settings">): IStateToProps => ({
  language: state.settings.language,
  translations: componentTranslationsSelector(state)
});

export const mapDispatchToProps = (dispatch: Dispatch): IDispatchToProps => {
  return {
    invokeChangeLanguage: (language: string) => dispatch(invokeChangeLanguage(language))
  };
};

const connectedAbout = connect<IStateToProps, IDispatchToProps>(mapStateToProps, mapDispatchToProps)(AboutPage);

export {AboutPage as UnconnectedAbout, connectedAbout as AboutPage};
