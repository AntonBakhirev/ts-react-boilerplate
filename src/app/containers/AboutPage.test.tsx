import {shallow} from "enzyme";
import * as React from "react";
import {IAction, IState} from "../redux/modules/baseModule";
import {ILanguage, SWITCH_LANGUAGE} from "../redux/modules/languageModule";
import {mapStateToProps, UnconnectedAbout} from "./AboutPage";

/* tslint:disable:no-empty jsx-no-lambda */
describe("<AboutPage />", () => {
  const language: IState<ILanguage> = {
    payload: {
      languageData: {greeting: "Hello!"},
      locale: "en-GB"
    }
  };

  it("matches snapshot", () => {
    const shallowComponent = shallow(<UnconnectedAbout dispatch={() => {}} locale=""/>);
    expect(shallowComponent).toMatchSnapshot();
  });

  it("maps state to props correctly", () => {
    const props = mapStateToProps({language});
    expect(props.locale).toEqual("en-GB");
  });

  it("calls switchLanguage() when button is clicked", () => {
    const spy = jest.spyOn(UnconnectedAbout.prototype, "switchLanguage");
    const shallowComponent = shallow(<UnconnectedAbout dispatch={() => {}} locale=""/>);

    expect(shallowComponent.find("button")).toBeDefined();
    expect(spy).not.toHaveBeenCalled();
    shallowComponent.find("button").simulate("click");
    expect(spy).toHaveBeenCalled();
  });

  describe("switchLanguage()", () => {
    it("dispatches SWITCH_LANGUAGE action to en-GB", () => {
      const dispatch = jest.fn();
      const shallowComponent = shallow(<UnconnectedAbout dispatch={dispatch} locale="" />);
      const expectedValue: IAction<ILanguage> = {
        payload: {
          languageData: {},
          locale: "en-GB"
        },
        type: SWITCH_LANGUAGE
      };

      expect(dispatch).not.toHaveBeenCalled();
      (shallowComponent as any).instance().switchLanguage();
      expect(dispatch).toHaveBeenCalledWith(expectedValue);
    });

    it("dispatches SWITCH_LANGUAGE action to de", () => {
      const dispatch = jest.fn();
      const shallowComponent = shallow(<UnconnectedAbout dispatch={dispatch} locale="en-GB"/>);
      const expectedValue: IAction<ILanguage> = {
        payload: {
          languageData: {},
          locale: "de"
        },
        type: SWITCH_LANGUAGE
      };

      expect(dispatch).not.toHaveBeenCalled();
      (shallowComponent as any).instance().switchLanguage();
      expect(dispatch).toHaveBeenCalledWith(expectedValue);
    });
  });

});
