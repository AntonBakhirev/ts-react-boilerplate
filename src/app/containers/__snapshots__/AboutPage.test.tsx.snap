// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AboutPage /> matches snapshot 1`] = `
<div>
  <h3>
    <FormattedMessage
      defaultMessage="Current Language"
      id="current.language"
      values={Object {}}
    />
    : 
  </h3>
  <button
    onClick={[Function]}
  >
    <FormattedMessage
      defaultMessage="Change language"
      id="about.change"
      values={Object {}}
    />
  </button>
  <h4>
    <FormattedMessage
      defaultMessage="About Us"
      id="about.us"
      values={Object {}}
    />
  </h4>
</div>
`;
